### PR TITLE
Simple Simulator: Created PhysicsBall abstraction

### LINKED ISSUE #869: Simple Simulator: Decide on physics library
### Description of the task

See #768 for more information about the simulator

Decide on the physics library to use for the simple simulator. Considerations:
- Available documentation
- 2D vs 3D
- General ease of use

### Acceptance criteria

- [ ] Decide on a physics library

### LINKED ISSUE #881: Simple Simulator: Create a simple PhysicsBall abstraction
### Description of the task

See #768 for more information on the simple simualator

Create a PhysicsBall class to make it easy to create a Ball object in the Box2D physics world and convert the data to our own `Ball` class.

### Acceptance criteria

- [x] Create a `PhysicsBall` class
- [x] Unit tests

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

This PR adds a `PhysicsBall` abstraction for the simulation, making it easy to add a Ball into the physics world as well as get a `Ball` object with the current ball state.

Box2D was chosen as the physics engine / library to use because of it's good documentation and overall simplicity.

Some utility functions for working with Box2D were added as well.

### Testing Done

unit tests

### Resolved Issues

resolves #881 
resolves #869 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — COMMENTED** [review]
Looks good, but why 2D and what's the plan for chipping?

File: src/software/backend/simulator_backend.h
Comment (garethellis0) [inline src/software/backend/simulator_backend.h]: TODO?

File: src/software/backend/simulation/physics_ball.cpp
Comment (garethellis0) [inline src/software/backend/simulation/physics_ball.cpp:32]: Where are we getting this constants from? Do they accurately reflect the real-world ball?

File: src/software/backend/simulation/physics_ball.cpp
Comment (garethellis0) [inline src/software/backend/simulation/physics_ball.cpp:14]: Perhaps link to the docs for what a "bullet" means?

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (garethellis0) [inline src/software/backend/simulation/physics_ball_test.cpp:125]: Should we add a test for bouncing/deflection?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, just some small doc comments and one test comment.

**Comment by MathewMacDougall** [comment]
@jonathanlew 2D is much simpler. The plan for chipping is to calculate the arc ourselves (or the points between which the ball is above the height of a robot) and disable collisions for that time to pretend the ball went over the robots.

File: src/software/backend/simulator_backend.h
Comment (MathewMacDougall) [reply src/software/backend/simulator_backend.h]: fixed

File: src/software/backend/simulation/physics_ball.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball.cpp:32]: These are untuned and somewhat arbitrary, but "ideal" values. Added a comment.

File: src/software/backend/simulation/physics_ball.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball.cpp:14]: done

**Comment by MathewMacDougall** [comment]
So a couple interesting (unfortunate) things I have discovered.
* Box2D defines a `velocityThreshold` below which collisions are treated as not elastic, and this value is 1. This means we either need to modify Box2D in some way to change this value (since it's a `#define`) or scale our values via our `PhysicsBall` abstraction. This SO post explains it pretty well: https://gamedev.stackexchange.com/questions/98851/body-lost-a-velocity-component-when-hitting-a-wall
* Box2D _also_ doesn't allow polygons with more than 8 vertices. This is much less of an issue than the velocity one, but makes properly approximating robot shaped harder.

I'm going to put this on the backburner for a few days while I focus on migrating old strategy code, but am open to solutions for the above problems, and might spend some time looking at physics engines again.

**Comment by garethellis0** [comment]
- Seems like scaling all the velocities (and spatial dimensions accordingly) is the only real solution for the first point, bit of a pain but manageable?
- Seems like we're just have to live with having a max of 8 points for representing robots. I don't think this would be a particularly large problem, if one at all.

**Comment by MathewMacDougall** [comment]
>     * Seems like scaling all the velocities (and spatial dimensions accordingly) is the only real solution for the first point, bit of a pain but manageable?
We could also fork Box2D and change the define there. Bazel makes that super easy. However this is a more decoupled solution, but does save the pain of converting units everywhere. Thoughts on this? I think I would prefer this option, at least for now. There are feature requests to Box2D to properly expose this parameter but who knows when that may come through.


>     * Seems like we're just have to live with having a max of 8 points for representing robots. I don't think this would be a particularly large problem, if one at all.
I guess just means we slightly under represent the collision area of a robot. In most cases shouldn't cause problems since we don't expect robots to collide and we can just check for "collision" in the tests just using the actual robot radius.

**Comment by jared-paul** [comment]
> If we fork it and change it, is it possible that it will mess with some internals of Box2D and cause problems in the future?

It seems like a safe modification given that it's just a threshold used to stop infinite collisions. I took a quick look at the code and online, and seems ok to change.

> Will we lose resolution if we scale the units down and back up to match Box2D's guideline?

I don't think so

> Also, probably a dumb question, but wouldn't we represent robots as circles, not polygons?

Ideally yes, but we really need the flat front of the robot to do kicking and dribbling, and there isn't a compound "arc" shape to only do part of the circle :cry:

**Comment by MathewMacDougall** [comment]
Whoops fat-fingered the close button.

Was doing some more reading and found that there's a rule of thumb to not have objects move more than 2m per time step (in our case 1/60 of a second) because this is really pushing the capabilities of the continuous-physics calculations and you will lose a lot of simulation accuracy (especially with collision) if things go faster. 

If we did scale our values to get around the velocity threshold, the ball could move up to 6500 units per second, which is ~108 units per 1/60 second. I think this would cause problems with the simulation, so I don't think scaling is an option.

**Comment by garethellis0** [comment]
Wut? You just need to scale your timestep proportionally no? In any case we will probably end up compiling box2d from source anyhow, so applying a quick patch before we do that should be easy enough. If that's the only issue, it's not show-stopper for me.

**Comment by MathewMacDougall** [comment]
Sure but your timestep would be 1/3250 of a second at best, which is basically useless because of how slow it is.

**Comment by garethellis0** [comment]
Is it slow? If it takes less then 1/3250 to actually do the math for that time step then we can run in real-time still.

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball_test.cpp:125]: Added tests for bouncing and deflection now.

**Comment by MathewMacDougall** [comment]
Hooray! I forked Box2D (our version is now here: https://github.com/UBC-Thunderbots/Box2D) and changed the constants, and everything works great now. I added 2 more unit tests for bouncing and deflection, so all should be good for final review now.

Please take a look at my commits to https://github.com/UBC-Thunderbots/Box2D as well and lmk if you have any comments / questions.

**Comment by garethellis0** [comment]
Don't suppose we could submit a PR to Box2D that allows compile-time configuration of the constants via whatever build system they use? This way we wouldn't have to worry about maintaining our own fork.

**Comment by garethellis0** [comment]
OH, we don't even build it eh? Could we just add something to our bazel build script to modify the constant?

**Comment by garethellis0** [comment]
Ok, just took a look at their code. Could we submit a PR to change `b2settings.h`:
```
#define b2_velocityThreshold		1.0f
```
to
```
#ifndef b2_velocityThreshold
#define b2_velocityThreshold		1.0f
#endif
```

Then in our bazel build file we could add `-Db2_velocityThreshold=0.2` (or whatever our threshold is) as a `cxxopt` when we're building the library.

**Comment by MathewMacDougall** [comment]
Unfortunately they don't accept PRs directly, it sounds like there's a longer issue-based process. I'll see what I can do but don't expect anything quick

**Comment by MathewMacDougall** [comment]
Is this going to be a blocker? There's no existing issue for the `b2Settings.h` file but it doesn't look like the developers are very active replying to the issues, so this won't be a fast process if it even happens?

**Comment by MathewMacDougall** [comment]
Alternatively, would you prefer my changes to our fork of Box2D be what you suggested with `#ifndef`?

**Comment by garethellis0** [comment]
let's:
- make the change on our fork
- open a PR from our fork to the main repo
- use our fork until that PR merges (or until the devs say no, in which case we should come up with something else)

**Comment by MathewMacDougall** [comment]
Ok I have updated our fork to use @garethellis0 's `#ifndef` solution, and updated the code to use it accordingly.

I have opened https://github.com/erincatto/Box2D/issues/559 with Box2D so we will see what happens.

I think this PR is good to go now.

**Comment by MathewMacDougall** [comment]
@garethellis0 @jared-paul please re-review

**Comment by garethellis0** [comment]
@MathewMacDougall - can you open a pull request from your fork? It's easier for the devs to accept a change if you hand them the code on a silver platter :stuck_out_tongue:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
@garethellis0 The contribution guide asked not to submit PRs directly and asked to open a ticket first for discussion

File: src/software/backend/simulation/physics_ball.cpp
Comment (jared-paul) [inline src/software/backend/simulation/physics_ball.cpp:51]: This might cause problems if we try to destroy a body mid simulation. I think Box2D locks the world during its timesteps so we can't edit it.

**Review by jared-paul — COMMENTED** [review]
Looks good! One small thing though

File: src/software/backend/simulation/physics_ball.cpp
Comment (garethellis0) [reply src/software/backend/simulation/physics_ball.cpp:51]: wow, nice catch.

Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball.cpp:51]: @jared-paul locks the works as-in with a mutex? Or just completely prevents changes from happening? I'd be surprised that it exposes this as an API if it wouldn't work. Can you explain more?

Comment (jared-paul) [reply src/software/backend/simulation/physics_ball.cpp:51]: I don't know how it locks it, I just know it does during timestep executions. For example, if you were to destroy a body in the middle of a collision that would be bad, so Box2D will throw a "world IsLocked" exception. This might not be the behaviour now, I just know that it used to be like that, they might've updated it since then.

https://www.iforce2d.net/b2dtut/removing-bodies here's a good source

Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball.cpp:51]: Ah ok I understand, good catch. For now this removal won't happen during a collision callback or during a physics step (anything that might need callbacks is TBD and needs to be set up). This should be safe for now but I'll definitely keep this in mind going forwards.

Sound good?

Comment (jared-paul) [reply src/software/backend/simulation/physics_ball.cpp:51]: Yep!

Quick thing: should we make a "PhysicsWorld" abstraction like we did with the ball? so instead of having "removeFromWorld" in "PhysicsBall" we could have "removeBall()" in "PhysicsWorld". I think it would remove the util class too

Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball.cpp:51]: I'll get there eventually, trying to build it up one class at a time so PRs aren't huge. Will definitely be revising stuff as I go.

Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball.cpp:51]: @jared-paul just need your approval then

**Review by jared-paul — APPROVED** [review]
LGTM! :+1: Can't wait to help out with this!

**Pull request merged**
